instance-scheduler/main.go

```diff
@@ -0,0 +1,285 @@
+		VersionStage: aws.String("AWSCURRENT"),
+	}
+	result, err := client.GetSecretValue(context.TODO(), input)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return *result.SecretString
+}
+
+func getNonProductionAccounts(cfg aws.Config, skipAccountNames string) map[string]string {
+	accounts := make(map[string]string)
+	// Get accounts secret
+	environments := getSecret(cfg, "environment_management")
+
+	var allAccounts map[string]interface{}
+	json.Unmarshal([]byte(environments), &allAccounts)
+
+	for _, record := range allAccounts {
+		if rec, ok := record.(map[string]interface{}); ok {
+			for key, val := range rec {
+				// Skip if the account's name ends with "-production", for example: performance-hub-production will be skipped
+				if !strings.HasSuffix(key, "-production") && (len(skipAccountNames) < 1 || !strings.Contains(skipAccountNames, key)) {
+					accounts[key] = val.(string)
+				}
+			}
+		}
+	}
+	return accounts
+}
+
+func stopInstance(client *ec2.Client, instanceId string) {
+	input := &ec2.StopInstancesInput{
+		InstanceIds: []string{
+			instanceId,
+		},
+		DryRun: aws.Bool(true),
+	}
+
+	_, err := client.StopInstances(context.TODO(), input)
+
+	var apiErr smithy.APIError
+	if errors.As(err, &apiErr) && apiErr.ErrorCode() == "DryRunOperation" {
+		log.Println("User has permission to stop instances.")
+		input.DryRun = aws.Bool(false)
+		_, err = client.StopInstances(context.TODO(), input)
+	}
+
+	if err == nil {
+		log.Printf("Successfully stopped instance with Id %v\n", instanceId)
+	} else {
+		log.Printf("ERROR: Could not stop instance: %v\n", err)
+	}
+}
+
+func startInstance(client *ec2.Client, instanceId string) {
+	input := &ec2.StartInstancesInput{
+		InstanceIds: []string{
+			instanceId,
+		},
+		DryRun: aws.Bool(true),
+	}
+
+	_, err := client.StartInstances(context.TODO(), input)
+
+	var apiErr smithy.APIError
+	if errors.As(err, &apiErr) && apiErr.ErrorCode() == "DryRunOperation" {
+		log.Println("User has permission to start an instance.")
+		input.DryRun = aws.Bool(false)
+		_, err = client.StartInstances(context.TODO(), input)
+	}
+
+	if err == nil {
+		log.Printf("Successfully started instance with Id %v\n", instanceId)
+	} else {
+		log.Printf("ERROR: Could not start instance: %v\n", err)
+	}
+}
+
+func stopStartInstancesInMemberAccount(client *ec2.Client, action string) {
+	input := &ec2.DescribeInstancesInput{}
+
+	result, err := client.DescribeInstances(context.TODO(), input)
+	if err != nil {
+		log.Print("ERROR: Could not retrieve information about Amazon EC2 instances in member account:\n", err)
+		return
+	}
+
+	instancesActedUpon := []string{}
+	skippedInstances := []string{}
+	skippedAutoScaledInstances := []string{}
+	for _, r := range result.Reservations {
+		for _, i := range r.Instances {
+			var instanceSchedulingTag string
+			instanceIsPartOfAutoScalingGroup := false
+			for _, tag := range i.Tags {
+				if *tag.Key == "instance-scheduling" {
+					instanceSchedulingTag = *tag.Value
```

Comment:
Sorry George I don't mean to nitpick, but I think these shouldn't be wrapped in an else. It's possible it could have an instance scheduling-tag and the autoscaling tag. So I'd just do them in in 2 if-blocks in that makes sense. Just in case users make a mistake or use some kind of common tag map for all their assets.

Reply:
If I am being honest, I am not sure I understand.

```
			for _, tag := range i.Tags {
				if *tag.Key == "aws:autoscaling:groupName" {
					instanceIsPartOfAutoScalingGroup = true
					break
				} else if *tag.Key == "instance-scheduling" {
					instanceSchedulingTag = *tag.Value
				}
			}
```

In the above, we iterate through the list of tags attached to an EC2 instance. I understand that an instance can have both the "instance-scheduling" and the "aws:autoscaling:groupName" tags associated with it. Every iteration inside the loop checks for the *tag.Key value. I am not sure if this is what you're suggesting, but I don't see how both conditions `*tag.Key == "aws:autoscaling:groupName"` and `*tag.Key == "instance-scheduling"` could be true at the same time, because they both check for the same value `*tag.Key` within the iteration. For example, I don't see any difference with using two ifs, in fact it would be slightly less efficient. Let me know what you think.

Reply:
It was only an issue if they were checked in the order of "does it have an instance-scheduling tag?" else "does it have an autoscaling tag." If you check for autoscaling first, it works great. I tried to say the same in the slack thread, sorry if that didn't come through.

`George, when reading the instance list, you check for autoscaling first then for scheduling. If you want to use if-else, when checking the tags, then doing it in that order instead would work just as well. :slightly_smiling_face:`